File: be/src/main/java/sg/flow/bootstrap/InjectTestData.java
package sg.flow.bootstrap;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.context.annotation.Profile;
import org.springframework.core.annotation.Order;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;
import sg.flow.entities.Account;
import sg.flow.entities.Bank;
import sg.flow.entities.Card;
import sg.flow.entities.TransactionHistory;
import sg.flow.entities.User;
import sg.flow.entities.utils.AccountType;
import sg.flow.entities.utils.CardType;
import sg.flow.repositories.account.AccountRepository;
import sg.flow.repositories.bank.BankRepository;
import sg.flow.repositories.card.CardRepository;
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository;
import sg.flow.repositories.user.UserRepository;

@RequiredArgsConstructor
@Profile("!prod")
@Component
@Order(2)
public class InjectTestData implements CommandLineRunner {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private AccountRepository accountRepository;

    @Autowired
    private CardRepository cardRepository;

    @Autowired
    private BankRepository bankRepository;

    @Autowired
    private TransactionHistoryRepository transactionRepository;

    @Override
    public void run(String... args) throws Exception {
        clearTestData();
        injectUserData();
        injectBankData();
        injectAccountData();
        injectCardData();
        injectTransactionData();
    }

    private void clearTestData() {
        transactionRepository.deleteAll();
        cardRepository.deleteAll();
        accountRepository.deleteAll();
        userRepository.deleteAll();
        bankRepository.deleteAll();
    }

    private BufferedReader readCSVFile(String csvFile) {
        BufferedReader br = null;
        try {
            BufferedReader bufferedReader = new BufferedReader(
                    new InputStreamReader(new ClassPathResource(csvFile).getInputStream()));
            br = bufferedReader;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return br;
    }

    private void injectUserData() {
        BufferedReader br = readCSVFile("testdata/users.csv");
        if (br == null) {
            System.out.println("Error reading user data file");
            return;
        }
        String line;
        try (br) {
            boolean isFirstLine = true;
            while ((line = br.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue;
                }
                String[] data = line.split(",");
                Integer id = Integer.parseInt(data[0]);
                String name = data[1];
                String email = data[2];
                String identificationNumber = data[3];
                String phoneNumber = data[4];
                LocalDate dateOfBirth = LocalDate.parse(data[5]);
                String address = data[6];
                String settingJson = data[7];

                User currentUser = User.builder()
                        .id(id)
                        .name(name)
                        .email(email)
                        .identificationNumber(identificationNumber)
                        .phoneNumber(phoneNumber)
                        .dateOfBirth(dateOfBirth)
                        .address(address)
                        .settingJson(settingJson)
                        .build();

                userRepository.save(currentUser);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void injectAccountData() {
        BufferedReader br = readCSVFile("testdata/accounts.csv");
        if (br == null) {
            System.out.println("Error reading account data file");
            return;
        }
        String line;
        try (br) {
            boolean isFirstLine = true;
            while ((line = br.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue;
                }
                String[] data = line.split(",");
                Long id = Long.parseLong(data[0]);
                String accountNumber = data[1];
                Long bankId = Long.parseLong(data[2]);
                Long ownerId = Long.parseLong(data[3]);
                Double balance = Double.parseDouble(data[4]);
                String accountName = data[5];
                String accountType = data[6];
                Double interestRatePerAnnum = Double.parseDouble(data[7]);
                LocalDateTime lastUpdated = LocalDateTime.parse(data[8]);

                Account currentAccount = Account.builder()
                        .id(id)
                        .accountNumber(accountNumber)
                        .bank(bankRepository.findById(bankId).get())
                        .owner(userRepository.findById(ownerId).get())
                        .balance(balance)
                        .accountName(accountName)
                        .accountType(AccountType.valueOf(accountType))
                        .interestRatePerAnnum(interestRatePerAnnum)
                        .lastUpdated(lastUpdated)
                        .build();

                accountRepository.save(currentAccount);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void injectCardData() {
        BufferedReader br = readCSVFile("testdata/cards.csv");
        if (br == null) {
            System.out.println("Error reading card data file");
            return;
        }
        String line;
        try (br) {
            boolean isFirstLine = true;
            while ((line = br.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue;
                }
                String[] data = line.split(",");
                Long id = Long.parseLong(data[0]);
                Long ownerId = Long.parseLong(data[1]);
                String cardNumber = data[2];
                Long issuingBankId = Long.parseLong(data[3]);
                Long linkedAccountId = Long.parseLong(data[4]);
                String cardType = data[5];
                LocalDate cardExpiryDate = LocalDate.parse(data[6]);
                String cardCvc = data[7];
                String cardPin = data[8];
                String cardStatus = data[9];
                String addressLine1 = data[10];
                String addressLine2 = data[11];
                String city = data[12];
                String state = data[13];
                String country = data[14];
                String zipCode = data[15];
                String phone = data[16];
                Double dailyLimit = Double.parseDouble(data[17]);
                Double monthlyLimit = Double.parseDouble(data[18]);
                String cardHolderName = data[19];

                Card currentCard = Card.builder()
                        .id(id)
                        .owner(userRepository.findById(ownerId).get())
                        .cardNumber(cardNumber)
                        .issuingBank(bankRepository.findById(issuingBankId).get())
                        .linkedAccount(accountRepository.findById(linkedAccountId).get())
                        .cardType(CardType.valueOf(cardType))
                        .cardHolderName(cardHolderName)
                        .expiryDate(cardExpiryDate)
                        .cvv(cardCvc)
                        .pin(cardPin)
                        .cardStatus(cardStatus)
                        .addressLine1(addressLine1)
                        .addressLine2(addressLine2)
                        .city(city)
                        .state(state)
                        .country(country)
                        .zipCode(zipCode)
                        .phone(phone)
                        .dailyLimit(dailyLimit)
                        .monthlyLimit(monthlyLimit)
                        .build();

                cardRepository.save(currentCard);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void injectBankData() {
        BufferedReader br = readCSVFile("testdata/banks.csv");
        if (br == null) {
            System.out.println("Error reading card data file");
            return;
        }
        String line;
        try (br) {
            boolean isFirstLine = true;
            while ((line = br.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue;
                }
                String[] data = line.split(",");
                Integer id = Integer.parseInt(data[0]);
                String name = data[1];
                String bankCode = data[2];

                Bank currentBank = Bank.builder()
                        .id(id)
                        .name(name)
                        .bankCode(bankCode)
                        .build();

                bankRepository.save(currentBank);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void injectTransactionData() {
        BufferedReader br = readCSVFile("testdata/transaction_history.csv");
        if (br == null) {
            System.out.println("Error reading card data file");
            return;
        }
        String line;
        try (br) {
            boolean isFirstLine = true;
            while ((line = br.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue;
                }
                String[] data = line.split(",");
                Long id = Long.parseLong(data[0]);
                Long toAccountId = Long.parseLong(data[1]);
                Long fromAccountId = Long.parseLong(data[2]);

                Long cardId;

                if (data[3].length() == 0) {
                    cardId = null;
                } else {
                    cardId = Long.parseLong(data[3]);
                }

                LocalDate transactionDate = LocalDate.parse(data[4]);
                LocalTime transactionTime = LocalTime.parse(data[5]);
                String description = data[6];
                Double amount = Double.parseDouble(data[7]);
                String transactionType = data[8];
                String transactionStatus = data[9];

                TransactionHistory currentTransaction = TransactionHistory.builder()
                        .id(id)
                        .toAccount(accountRepository.findById(toAccountId).orElse(null))
                        .fromAccount(accountRepository.findById(fromAccountId).orElse(null))
                        .card(cardRepository.findById(cardId).orElse(null))
                        .transactionDate(transactionDate)
                        .transactionTime(transactionTime)
                        .description(description)
                        .amount(amount)
                        .transactionType(transactionType)
                        .transactionStatus(transactionStatus)
                        .build();

                transactionRepository.save(currentTransaction);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: be/src/main/kotlin/sg/flow/bootstrap/InjectTestData.kt
package sg.flow.bootstrap

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import kotlinx.coroutines.*
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.CommandLineRunner
import org.springframework.context.annotation.Profile
import org.springframework.core.annotation.Order
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component
import sg.flow.entities.Account
import sg.flow.entities.Bank
import sg.flow.entities.Card
import sg.flow.entities.TransactionHistory
import sg.flow.entities.User
import sg.flow.entities.utils.AccountType
import sg.flow.entities.utils.CardType
import sg.flow.models.card.BriefCard
import sg.flow.repositories.account.AccountRepository
import sg.flow.repositories.bank.BankRepository
import sg.flow.repositories.card.CardRepository
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository
import sg.flow.repositories.user.UserRepository
import org.slf4j.LoggerFactory

@Profile("!prod")
@Component
@Order(2)
class InjectTestData(
        @Autowired private val userRepository: UserRepository,
        @Autowired private val accountRepository: AccountRepository,
        @Autowired private val cardRepository: CardRepository,
        @Autowired private val bankRepository: BankRepository,
        @Autowired private val transactionRepository: TransactionHistoryRepository
) : CommandLineRunner {

    private val logger = LoggerFactory.getLogger(InjectTestData::class.java)

    override fun run(vararg args: String) = runBlocking {
        logger.info("Starting test data injection...")

//        clearTestData()
//        logger.info("Cleared existing test data.")

//        injectUserData()
        logger.info("Injected user data.")

//        injectBankData()
//        logger.info("Injected bank data.")

//        injectAccountData()
//        logger.info("Injected account data.")
//
//        injectCardData()
//        logger.info("Injected card data.")
//
//        injectTransactionData()
//        logger.info("Injected transaction data.")

        logger.info("Test data injection completed successfully.")
    }

    private suspend fun clearTestData() {
        transactionRepository.deleteAll()
        cardRepository.deleteAll()
        accountRepository.deleteAll()
        userRepository.deleteAll()
        bankRepository.deleteAll()
    }

    private fun readCSVFile(csvFile: String): BufferedReader? {
        return try {
            BufferedReader(InputStreamReader(ClassPathResource(csvFile).inputStream))
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    private suspend fun injectUserData() {
        val br = readCSVFile("testdata/users.csv")
        if (br == null) {
            logger.error("Error reading user data file")
            return
        }

        br.use { reader ->
            var isFirstLine = true
            reader.lineSequence().forEach { line ->
                if (isFirstLine) {
                    isFirstLine = false
                    return@forEach
                }

                val data = line.split(",")
                val currentUser =
                        User(
                                id = data[0].toInt(),
                                name = data[1],
                                email = data[2],
                                identificationNumber = data[3],
                                phoneNumber = data[4],
                                dateOfBirth = LocalDate.parse(data[5]),
                                address = data[6],
                                gender_is_male = null,
                                settingJson = data[7]
                        )

                userRepository.save(currentUser)
            }
        }
    }

    private suspend fun injectAccountData() {
        val br = readCSVFile("testdata/accounts.csv")
        if (br == null) {
            logger.error("Error reading account data file")
            return
        }

        br.use { reader ->
            var isFirstLine = true
            reader.lineSequence().forEach { line ->
                if (isFirstLine) {
                    isFirstLine = false
                    return@forEach
                }

                val data = line.split(",")
                val currentAccount =
                        Account(
                                id = data[0].toLong(),
                                accountNumber = data[1],
                                bank = bankRepository.findById(data[2].toLong())!!,
                                owner = userRepository.findById(data[3].toLong())!!,
                                balance = data[4].toDouble(),
                                accountName = data[5],
                                accountType = AccountType.valueOf(data[6]),
                                interestRatePerAnnum = data[7].toDouble(),
                                lastUpdated = LocalDateTime.parse(data[8]),
                                finverseId = null,
                        )

                accountRepository.save(currentAccount)
            }
        }
    }

    private suspend fun injectCardData() {
        val br = readCSVFile("testdata/cards.csv")
        if (br == null) {
            logger.error("Error reading card data file")
            return
        }

        br.use { reader ->
            var isFirstLine = true
            reader.lineSequence().forEach { line ->
                if (isFirstLine) {
                    isFirstLine = false
                    return@forEach
                }

                val data = line.split(",")
                val currentCard =
                        Card(
                                id = data[0].toLong(),
                                owner = userRepository.findById(data[1].toLong())!!,
                                cardNumber = data[2],
                                issuingBank = bankRepository.findById(data[3].toLong())!!,
                                cardType = CardType.valueOf(data[5]),
                        )

                cardRepository.save(currentCard)
            }
        }
    }

    private suspend fun injectBankData() {
        val br = readCSVFile("testdata/banks.csv")
        if (br == null) {
            logger.error("Error reading bank data file")
            return
        }

        br.use { reader ->
            var isFirstLine = true
            reader.lineSequence().forEach { line ->
                if (isFirstLine) {
                    isFirstLine = false
                    return@forEach
                }

                val data = line.split(",")
                val currentBank = Bank(id = data[0].toInt(), name = data[1], bankCode = data[2])

                bankRepository.save(currentBank)
            }
        }
    }

    private suspend fun injectTransactionData(
        batchSize: Int = 500,
        parallelism: Int = 20               // active parseLine coroutines at once
    ) = coroutineScope {

        val br       = readCSVFile("testdata/transaction_history.csv") ?: return@coroutineScope
        val gate     = Semaphore(parallelism)                           // the global “traffic‑light”

        br.useLines { lines ->
            lines
                .drop(1)                           // skip header
                .chunked(batchSize)                // Sequence<List<String>>
                .forEach { chunk ->                // process batches sequentially
                    // ── parse every row concurrently, but capped by gate ──
                    val entities = chunk.map { row ->
                        async {
                            gate.withPermit {       // <= only `parallelism` concurrent connections
                                parseLine(row, accountRepository, cardRepository)
                            }
                        }
                    }.awaitAll()                    // wait for this batch to finish parsing

                    transactionRepository.saveAllWithId(entities)
                }
        }
    }


    /**
     * Convert one CSV row into a fully–populated TransactionHistory.
     * Uses AccountRepository and CardRepository to hydrate references.
     */
    private suspend fun parseLine(
        line: String,
        accountRepository: AccountRepository,
        cardRepository: CardRepository
    ): TransactionHistory {

        val data = line.split(',')

        val cardId  = data[3].takeIf { it.isNotBlank() }?.toLong()
        val account = accountRepository.findById(data[2].toLong())

        val briefCard = cardId?.let { id ->
            cardRepository.findById(id)?.let { full ->
                BriefCard(full.id ?: -1, full.cardNumber, full.cardType)
            }
        }

        return TransactionHistory(
            id                   = data[0].toLong(),
            transactionReference = data[1],
            account              = account,
            card                 = briefCard,
            transactionDate      = LocalDate.parse(data[4]),
            transactionTime      = data[5].takeIf { it.isNotBlank() }?.let { LocalTime.parse(it) },
            amount               = data[6].toDouble(),
            transactionType      = data[7],
            description          = data[8],
            transactionStatus    = data[9],
            friendlyDescription  = data[10],
            finverseId = generateRandomString(255)
        )
    }

    private fun generateRandomString(length: Int = 255): String {
        val allowedChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return (1..length)
            .map { allowedChars.random() }
            .joinToString("")
    }

}
